### PR TITLE
2.x RetryableConnection for SenderReplication

### PR DESCRIPTION
**Pull request by qiangdavidliu**
Using RetryableConnection for sender replication.
Connecting connection lifecycle to channel lifecycle in AbstractClientChannel.

### CODE REVIEW

File: eureka2-core/src/main/java/com/netflix/eureka2/connection/RetryableConnectionFactory.java [hunk 1]
```diff
@@ -14,50 +14,55 @@
...
     public RetryableConnectionFactory(final ChannelFactory<CHANNEL> channelFactory) {
         this.channelFactory = channelFactory;
     }
 
     /**
+     * @param executeOnChannel a func1 that describes a nullary call for the channel
+     * @return a {@link RetryableConnection} that contains several observables. This lifecycle observable provided
+     * can be retried on.
+     */
+    public RetryableConnection<CHANNEL> nullaryConnection(final Func1<CHANNEL, Observable<Void>> executeOnChannel) {
```
Comment (tbak) [inline eureka2-core/src/main/java/com/netflix/eureka2/connection/RetryableConnectionFactory.java]: I understand that this is type of connection where a client does not provide any operations. Can we pickup a better name for it, and also for unaryConnection? Some documentation would be useful, as well.

File: eureka2-core/src/main/java/com/netflix/eureka2/connection/RetryableConnectionFactory.java [hunk 2]
```diff
@@ -124,5 +129,23 @@ public void call(CHANNEL channel) {
         );
     }
 
-    protected abstract Observable<Void> executeOnChannel(CHANNEL channel, OP op);
```
Comment (tbak) [inline eureka2-core/src/main/java/com/netflix/eureka2/connection/RetryableConnectionFactory.java]: I have added in my branch 
```java
protected abstract Observable<INPUT> connectToChannelInput(CHANNEL channel);
```
How can we expose channel input with this implementation?

File: eureka2-write-server/src/test/java/com/netflix/eureka2/server/service/replication/ReplicationHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+        MessageConnection messageConnection = newMockMessageConnection();
+        when(messageConnection.submit(anyObject())).thenReturn(Observable.<Void>empty());
+        when(messageConnection.incoming()).thenReturn(Observable.<Object>just(HELLO_REPLY));
+
+        TransportClient transportClient = mock(TransportClient.class);
+        when(transportClient.connect()).thenReturn(Observable.just(messageConnection));
+
+        final ReplicationChannel channel = new SenderReplicationChannel(transportClient, mock(ReplicationChannelMetrics.class));
+
+        Observable.empty().delay(failAfterMillis, TimeUnit.MILLISECONDS).subscribe(new Subscriber<Object>() {
```
Comment (tbak) [inline eureka2-write-server/src/test/java/com/netflix/eureka2/server/service/replication/ReplicationHandlerTest.java]: Is it real time, or from TestScheduler? Why not schedule this as a task?

File: eureka2-core/src/main/java/com/netflix/eureka2/connection/RetryableConnectionFactory.java (hunk 1 above)
Comment (qiangdavidliu) [inline eureka2-core/src/main/java/com/netflix/eureka2/connection/RetryableConnectionFactory.java]: nullary and unary are pretty standard mathematical names for operand sizes: http://en.wikipedia.org/wiki/Arity.

File: eureka2-write-server/src/test/java/com/netflix/eureka2/server/service/replication/ReplicationHandlerTest.java (hunk 3 above)
Comment (qiangdavidliu) [inline eureka2-write-server/src/test/java/com/netflix/eureka2/server/service/replication/ReplicationHandlerTest.java]: This is intended as a realtime task as it need to sync with some other steps in the test. It'll be difficult to do as a testScheduler unfortunately.

File: eureka2-core/src/main/java/com/netflix/eureka2/connection/RetryableConnectionFactory.java (hunk 2 above)
Comment (qiangdavidliu) [inline eureka2-core/src/main/java/com/netflix/eureka2/connection/RetryableConnectionFactory.java]: the abstract executeOnChannel() functionality is now provide as a func1/func2 that are passed in, so RetryableConnectionFactory is no longer abstract.

File: eureka2-core/src/main/java/com/netflix/eureka2/connection/RetryableConnectionFactory.java (hunk 1 above)
Comment (tbak) [inline eureka2-core/src/main/java/com/netflix/eureka2/connection/RetryableConnectionFactory.java]: I understand that. But in nullaryConnection method name, it implies that the connection itself is nullary, and we actually mean here the operator. Maybe connectionWithNullaryOp?

File: eureka2-core/src/main/java/com/netflix/eureka2/connection/RetryableConnectionFactory.java (hunk 2 above)
Comment (tbak) [inline eureka2-core/src/main/java/com/netflix/eureka2/connection/RetryableConnectionFactory.java]: That does not answer the question, how we can now pass channel input. Shall I keep the abstract method or add another function argument. For consistency the latter is better, but the number of arguments becomes too big. We might be force to introduce a builder for more clarity.

**Pull request merged**